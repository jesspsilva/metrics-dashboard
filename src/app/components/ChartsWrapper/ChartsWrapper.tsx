import { useMemo } from "react";
import {
  formatNumber,
  fromHoursToMinutes,
  fromSecsToMinutes,
} from "@/app/utils/format-number";
import { getCategoryColor } from "@/app/utils/categories-colors";
import { filterDataByCategory } from "@/app/utils/filter-data-by-category";

import DonutChart from "@components/DonutChart/DonutChart";
import ProgressCircle from "@components/ProgressCircle/ProgressCircle";
import BarChart from "@/app/components/BarChart/BarChart";

export default function ChartsWrapper({
  data,
  category,
  onChange,
}: {
  data: MetricsData[];
  category: string;
  onChange: (v: string | number, type?: FiltersType) => void;
}) {
  const downtimeData = useMemo(
    () =>
      data
        ? data
            .filter((item) => item.category === "downtime")
            .map((item, index) => ({
              name: item.label,
              value: item.value,
              type: item.type,
              color: `${getCategoryColor("downtime").baseColor}-${200 + index * 600}`,
            }))
        : [],
    [data],
  );

  const categoryCounts = useMemo(() => {
    const count = {} as Record<string, number>;

    data.forEach((item) => {
      const category = item.category;
      if (count[category]) {
        count[category]++;
      } else {
        count[category] = 1;
      }
    });

    return Object.entries(count).map(([category, itemCount]) => ({
      name: category,
      value: itemCount,
      color: getCategoryColor(category).chartColor,
    }));
  }, [data]);

  const equipmentEfficiency = data.filter((item) => item.label === "oee")[0]
    .value;

  const efficiencyData = useMemo(
    () =>
      data
        .filter(
          (item) => item.category === "efficiency" && item.label !== "oee",
        )
        .map((item) => ({
          name: item.label,
          "Loss Value": formatNumber(item.value, item.type),
        })),
    [data],
  );

  const shiftData = useMemo(() => filterDataByCategory(data, "shift"), [data]);

  const shiftChartData = useMemo(() => {
    const shiftDuration =
      shiftData.find((item) =>
        item.name.toLowerCase().includes("shift duration"),
      )?.value || 0;

    const cleaningTime =
      shiftData.find((item) => item.name.toLowerCase().includes("cleaning"))
        ?.value || 0;

    const shiftHoursToMinutes = fromHoursToMinutes(shiftDuration);
    const cleaningSecondsToMinutes = fromSecsToMinutes(cleaningTime);

    return [
      {
        name: "Cleaning",
        value: cleaningSecondsToMinutes,
        color: `${getCategoryColor("shift").baseColor}-200`,
      },
      {
        name: "Other tasks",
        value: shiftHoursToMinutes - cleaningSecondsToMinutes,
        color: `${getCategoryColor("shift").baseColor}-800`,
      },
    ];
  }, [shiftData]);

  return (
    <>
      {category.toLowerCase() === "all" && (
        <div data-testid="all-chart" className="chart">
          <DonutChart
            variant="pie"
            data={categoryCounts}
            colors={categoryCounts.map((item) => item.color)}
            onChange={(value) => onChange(value, "categories")}
            title="Categories Distribution"
          />
        </div>
      )}

      {category.toLowerCase() === "downtime" && (
        <div data-testid="downtime-chart" className="chart">
          <DonutChart
            data={downtimeData}
            onChange={onChange}
            type={downtimeData[0].type}
            title="Downtime Distribution"
            colors={downtimeData.map((item) => item.color)}
          />
        </div>
      )}
      {category.toLowerCase() === "efficiency" && (
        <div
          data-testid="efficiency-charts"
          className="chart efficiency-charts"
        >
          <ProgressCircle
            value={equipmentEfficiency * 100}
            title="Equipment Efficiency"
            color={getCategoryColor("efficiency").baseColor}
          />
          <BarChart
            data={efficiencyData}
            title="Loss parameters distribution"
            onChange={onChange}
            categories={["Loss Value"]}
            colors={[getCategoryColor("efficiency").baseColor]}
          />
        </div>
      )}
      {category.toLowerCase() === "shift" && (
        <div data-testid="shift-chart" className="chart">
          <DonutChart
            data={shiftChartData}
            onChange={(value) => {
              if (value === "Other tasks") {
                onChange("shift", "categories");
                return;
              }

              onChange(value);
            }}
            type="minutes"
            title="Distribution of tasks during the shift"
            colors={shiftChartData.map((item) => item.color)}
          />
        </div>
      )}
    </>
  );
}
