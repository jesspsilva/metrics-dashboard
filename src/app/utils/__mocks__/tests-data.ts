export const metricsData = [
  {
    id: "oee",
    label: "oee",
    value: 0.68,
    type: "percentage",
    description: "The overall equipment efficiency in %",
    category: "efficiency",
  },
  {
    id: "sl",
    label: "Speed loss",
    value: -10.5,
    type: "number",
    description: "The line speed loss",
    category: "efficiency",
  },
  {
    id: "cln_shift",
    label: "Cleaning in shift",
    value: 2280,
    type: "secs",
    description:
      "Time spent (in seconds) cleaning the machines during the last shift",
    category: "shift",
  },
  {
    id: "shift_duration",
    label: "Shift duration",
    value: 8,
    type: "hours",
    description: "Last shift duration in hours",
    category: "shift",
  },
  {
    id: "no_value",
    label: "",
    value: null,
    type: "",
    description: "",
    category: "",
  },
  {
    id: "unexplained",
    label: "Unexplained downtime",
    value: 180,
    type: "secs",
    description: "The downtime (in seconds) caused by undocumented reasons",
    category: "downtime",
  },
  {
    id: "mech_problems",
    label: "Mechanical problems",
    value: 1210,
    type: "secs",
    description: "The downtime (in seconds) caused by mechanical problems",
    category: "downtime",
  },
] as MetricsData[];
