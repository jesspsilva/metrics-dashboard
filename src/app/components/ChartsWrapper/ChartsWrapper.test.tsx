import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import ChartsWrapper from "./ChartsWrapper";
import { metricsData as data } from "@/app/utils/__mocks__/tests-data";

// Added this mock because the ResponsiveContainer component from recharts
// is causing an error when running the tests
// Reference: https://github.com/recharts/recharts/issues/727
jest.mock("recharts", () => ({
  ...jest.requireActual("recharts"),
  ResponsiveContainer: jest
    .fn()
    .mockImplementation((props: any) => <div {...props} />),
}));

const defaultTestProps = {
  data,
  category: "efficiency",
  onChange: () => {},
};

const renderComponent = (props = {}) => {
  const mergedProps = { ...defaultTestProps, ...props };
  return render(<ChartsWrapper {...mergedProps} />);
};

describe("ChartsWrapper", () => {
  it("should render correctly", () => {
    const { asFragment } = renderComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  describe("when the category is downtime", () => {
    it("should render the correct downtime data", async () => {
      renderComponent({ category: "downtime" });

      const donutChart = await screen.findByTestId("downtime-chart");

      expect(donutChart).toBeInTheDocument();
      expect(donutChart).toHaveTextContent("Downtime Distribution");
      expect(donutChart?.children).toHaveLength(1);
    });
  });

  describe("when the category is efficiency", () => {
    it("should render the correct efficiency data", async () => {
      renderComponent({ category: "efficiency" });

      const efficiencyCharts = await screen.findByTestId("efficiency-charts");

      expect(efficiencyCharts?.children).toHaveLength(2);
    });
  });

  describe("when the category is shift", () => {
    it("should render the correct shift data", async () => {
      renderComponent({ category: "shift" });

      const donutChart = await screen.findByTestId("shift-chart");

      expect(donutChart).toBeInTheDocument();
      expect(donutChart).toHaveTextContent(
        "Distribution of tasks during the shift",
      );
      expect(donutChart?.children).toHaveLength(1);
    });
  });

  describe("when the category is all", () => {
    it("should render all categories chart data", async () => {
      renderComponent({ category: "All" });

      const charts = await screen.findByTestId("all-chart");

      expect(charts?.children).toHaveLength(1);
    });
  });
});
