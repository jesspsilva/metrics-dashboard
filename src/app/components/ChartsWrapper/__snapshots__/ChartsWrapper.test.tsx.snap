// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartsWrapper should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chart efficiency-charts"
    data-testid="efficiency-charts"
  >
    <div
      class="sc-bc09e72-0 kPbMxn"
    >
      <h2
        class="sc-bc09e72-1 gDcwJS"
      >
        Equipment Efficiency
      </h2>
      <div
        class="tremor-ProgressBar-root flex flex-col items-center justify-center"
      >
        <svg
          class="transform -rotate-90"
          height="160"
          viewBox="0 0 160 160"
          width="160"
        >
          <circle
            class="transition-colors ease-linear stroke-emerald-500 dark:stroke-emerald-500 opacity-20 dark:opacity-25"
            cx="80"
            cy="80"
            fill="transparent"
            r="73"
            stroke=""
            stroke-linecap="round"
            stroke-width="14"
          />
          <circle
            class="stroke-emerald-500 dark:stroke-emerald-500 transition-all duration-300 ease-in-out"
            cx="80"
            cy="80"
            fill="transparent"
            r="73"
            stroke=""
            stroke-dasharray="458.6725274241098 458.6725274241098"
            stroke-dashoffset="146.7752087757151"
            stroke-linecap="round"
            stroke-width="14"
          />
        </svg>
        <div
          class="absolute flex"
        >
          <span
            class="fill-tremor-content-emphasis"
          >
            68%
          </span>
        </div>
      </div>
    </div>
    <div
      class="sc-b143385c-0 jNqabe"
    >
      <h2
        class="sc-b143385c-1 mKzdD"
      >
        Loss parameters distribution
      </h2>
      <div
        class="w-full h-80 mt-6"
      >
        <div
          class="h-full w-full"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
